File: go/days/day01/part2.go
package day01

import (
	"bytes"
)

func extractValsPart2(lineBytes []byte) int {
	var (
		foundLower, foundUpper bool
		lower, upper           int
	)

	headStart, tailEnd := 0, len(lineBytes)
	headEnd, tailStart := 0, len(lineBytes)-1

	for headStart <= tailStart {
		if foundLower && foundUpper {
			break
		}

		if !foundLower {
			foundDigit, n := isDigit(lineBytes[headEnd])
			if foundDigit {
				lower = n
				foundLower = true
			}

			foundStrDigit, n := isStringDigit(lineBytes[headStart:headEnd])
			if foundStrDigit {
				lower = n
				foundLower = true
			}
			headEnd++
		}

		if !foundUpper {
			foundDigit, n := isDigit(lineBytes[tailStart])
			if foundDigit {
				upper = n
				foundUpper = true
			}

			foundStrDigit, n := isStringDigit(lineBytes[tailStart:tailEnd])
			if foundStrDigit {
				upper = n
				foundUpper = true
			}
			tailStart--
		}
	}

	if !foundLower {
		return upper*10 + upper
	}

	if !foundUpper {
		return lower*10 + lower
	}

	return lower*10 + upper
}

func isStringDigit(b []byte) (bool, int) {
	switch {
	case bytes.Contains(b, []byte("one")):
		return true, 1
	case bytes.Contains(b, []byte("two")):
		return true, 2
	case bytes.Contains(b, []byte("three")):
		return true, 3
	case bytes.Contains(b, []byte("four")):
		return true, 4
	case bytes.Contains(b, []byte("five")):
		return true, 5
	case bytes.Contains(b, []byte("six")):
		return true, 6
	case bytes.Contains(b, []byte("seven")):
		return true, 7
	case bytes.Contains(b, []byte("eight")):
		return true, 8
	case bytes.Contains(b, []byte("nine")):
		return true, 9
	default:
		return false, -1
	}
}

File: zig/days/day01/part2.zig
const std = @import("std");
const part1 = @import("./part1.zig");

pub fn extract_vals(line: []u8) u8 {
    var foundLower = false;
    var foundUpper = false;
    var lower: ?u8 = null;
    var upper: ?u8 = null;

    var headStart: usize = 0;
    var headEnd: usize = 0;
    var tailStart = line.len - 1;
    var tailEnd = line.len;

    while (headStart < tailStart) {
        if (foundLower and foundUpper) {
            break;
        }

        if (!foundLower) {
            const lookup = part1.isDigit(line[headEnd]);
            if (lookup.found) {
                lower = lookup.num;
                foundLower = true;
            }

            const strLookup = isStringDigit(line[headStart..headEnd]);
            if (strLookup.found) {
                lower = strLookup.num;
                foundLower = true;
            }

            headEnd += 1;
        }

        if (!foundUpper) {
            const lookup = part1.isDigit(line[tailStart]);
            if (lookup.found) {
                upper = lookup.num;
                foundUpper = true;
            }

            const strLookup = isStringDigit(line[tailStart..tailEnd]);
            if (strLookup.found) {
                upper = strLookup.num;
                foundUpper = true;
            }

            tailStart -= 1;
        }
    }

    if (!foundUpper) {
        return lower.? * 10 + lower.?;
    }

    if (!foundLower) {
        return upper.? * 10 + upper.?;
    }

    return lower.? * 10 + upper.?;
}

const isStringDigitResult = struct {
    found: bool,
    num: u8,
};

fn isStringDigit(b: []u8) isStringDigitResult {
    if (std.mem.containsAtLeast(u8, b, 1, "one")) {
        return isStringDigitResult{ .found = true, .num = 1 };
    }
    if (std.mem.containsAtLeast(u8, b, 1, "two")) {
        return isStringDigitResult{ .found = true, .num = 2 };
    }
    if (std.mem.containsAtLeast(u8, b, 1, "three")) {
        return isStringDigitResult{ .found = true, .num = 3 };
    }
    if (std.mem.containsAtLeast(u8, b, 1, "four")) {
        return isStringDigitResult{ .found = true, .num = 4 };
    }
    if (std.mem.containsAtLeast(u8, b, 1, "five")) {
        return isStringDigitResult{ .found = true, .num = 5 };
    }
    if (std.mem.containsAtLeast(u8, b, 1, "six")) {
        return isStringDigitResult{ .found = true, .num = 6 };
    }
    if (std.mem.containsAtLeast(u8, b, 1, "seven")) {
        return isStringDigitResult{ .found = true, .num = 7 };
    }
    if (std.mem.containsAtLeast(u8, b, 1, "eight")) {
        return isStringDigitResult{ .found = true, .num = 8 };
    }
    if (std.mem.containsAtLeast(u8, b, 1, "nine")) {
        return isStringDigitResult{ .found = true, .num = 9 };
    }

    return isStringDigitResult{ .found = false, .num = 0 };
}
